torch/_refs/__init__.py

```diff
@@ -5602,6 +5602,7 @@ def triu(a: TensorLikeType, diagonal: int = 0) -> TensorLikeType:
         torch.arange(w, device=a.device).unsqueeze(-2)
         - torch.arange(h, device=a.device).unsqueeze(-1)
     ) >= diagonal
+    mask = mask.broadcast_to(a.shape)
```

Comment:
Do you have a repro for this? I'm not really sure what the problem is

Reply:
Sure,

```
import torch

x = torch.ones(1, 12, 10, 10).to(torch.bool)
y = torch.ones(10, 10).to(torch.bool)
torch._refs.triu(x, y)
```

Then you see the warning.

In the future version when this warning becomes an Error, this decomp is no longer valid.

Reply:
The real issue here is that diagonal should be an int. If anything, this ref what's missing is a type check.

Reply:
But note that, in refs, we don't do a very exhaustive typecheck of the inputs (the annotations of the function are correct) so I think that this is a bit of a non-issue as it stands, since you can't hit this path when executing the decomposition.

Reply:
I did hit this issue on decomposition. The original model code is something along the lines of `a.triu(..)` and a is a bool tensor.

Are you suggesting that we should cast bool tensor to int tensor inside of this function?

Reply:
Can you provide a repro for how to hit this via a decomposition?

Reply:
Sure, I can try, it probably a long code.

However, `In the future version when this warning becomes an Error, this decomp is no longer valid.` <--- isn't this statement still a true statement regardless how you hit it?

Reply:
Hi @lezcano here is a repro: https://gist.github.com/qihqi/f75d205763160700374c72746eeee4ef

Thanks!

Reply:
>isn't this statement still a true statement regardless how you hit it?

Yes, but my point is that in the OP you are passing a Tensor where you should have an integer. See how it breaks in eager:
```python
>>> import torch
>>> 
>>> x = torch.ones(1, 12, 10, 10).to(torch.bool)
>>> y = torch.ones(10, 10).to(torch.bool)
>>> torch.triu(x, y)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: triu(): argument 'diagonal' (position 2) must be int, not Tensor
```

For the repro, please find how is is it that you are passing an invalid argument to `triu`, there is where the real bug is.

Reply:
Right, the second arg should be int; still reproducible:

```
>>> import torch
>>> 
>>> x = torch.ones(1, 12, 10, 10).to(torch.bool)
>>> y = torch.ones(10, 10).to(torch.bool)
>>> torch.triu(x, y)
... # no warning
>>> torch._refs.triu(x, 1)
/mnt/hanq/git/qihqi/pytorch/torch/_prims_common/__init__.py:1814: UserWarning: An output with one or more elements was resized since it had shape [10, 10], which does not match the required output shape [1, 12, 10, 10]. This behavior is deprecated, and in a future PyTorch release outputs will not be resized unless they have zero elements. You can explicitly reuse an out tensor t by resizing it, inplace, to zero elements with t.resize_(0). (Triggered internally at /mnt/hanq/git/qihqi/pytorch/aten/src/ATen/native/Resize.cpp:35.)
  return mask.logical_and(t)
```

Reply:
Fair enough. That being said, as the error shows, this is an issue with `logical_and` not supporting broadcasting properly. Feel free to send a PR fixing this.

Reply:
actually, no, that's not the case as I can't repro calling `_refs.logical_and. Not sure what's the issue, but it's somewhere in the stack. In particular, broadcasting  should happen automatically without any warning.

Reply:
So the q is, should `logical_and` support broadcast of 2 tensors with different ranks?

From the warning message, it seems to me that, it used to be support this case but is moving to deprecate that behavior. Is my reading correct?

Reply:
Not sure if related, but https://github.com/pytorch/pytorch/pull/112170 fixes a regression with logical_and